include/ppx/grfx/grfx_foveation.h

```diff
@@ -0,0 +1,134 @@
+
+#include "ppx/grfx/grfx_config.h"
+#include "ppx/grfx/grfx_texture.h"
+
+namespace ppx {
+namespace grfx {
+
+//! @struct FoveationCapabilities
+//!
+//!
+struct FoveationCapabilities
+{
+    struct
+    {
+        bool supported                                 = false;
+        bool supportsDynamicImageView                  = false;
+        bool supportsNonSubsampledImages               = false;
+        bool supportsAdditionalInvocations             = false;
+        bool supportsDeferred                          = false;
+        bool subsampledLoads                           = false;
+        bool subsampledCoarseReconstructionEarlyAccess = false;
+        struct
+        {
+            struct
+            {
+                uint32_t width;
+                uint32_t height;
+            } min;
+            struct
+            {
+                uint32_t width;
+                uint32_t height;
+            } max;
+        } texelSize;
+        uint32_t maxSubsampledArrayLayers;
+        uint32_t maxDescriptorSetSubsampledSamplers;
+    } fdm;
+
+    struct
+    {
+        bool  supportPipelineVRS   = false;
+        bool  supportPrimitiveVRS  = false;
+        bool  supportAttachmentVRS = false;
+        uint2 minTexelSize;
+        uint2 maxTexelSize;
+        struct
+        {
+            // List of supported encoded shading rates
+            uint32_t supportedRateCount = 0;
+            uint8_t  supportedRates[11];
+
+            // mapRateToSpported[shadingRate] is a supported shading rate whose
+            // fragment size that is no larger than `shadingRate` in either
+            // dimension.
+            uint8_t mapRateToSupported[11];
+
+        } vk;
+    } vrs;
+};
+
+//! @struct FoveationPatternCreateInfo
+//!
+//!
+struct FoveationPatternCreateInfo
+{
+    uint32_t             fbWidth       = 0;
+    uint32_t             fbHeight      = 0;
+    FoveationMode        foveationMode = FOVEATION_NONE;
+    FoveationPatternType patternType   = FOVEATION_PATTERN_TYPE_UNIFORM_1X1;
+};
+
+//! @class FoveationPattern
+//!
+//!
+class FoveationPattern
+    : public DeviceObject<FoveationPatternCreateInfo>
+{
+public:
+    FoveationPattern() {}
+    virtual ~FoveationPattern() {}
+
+    FoveationMode       GetFoveationMode() const { return mFoveationMode; }
+    SampledImageViewPtr GetFoveationImageViewPtr() const { return mFoveationTexture->GetSampledImageView(); }
+    uint32_t            GetFoveationTextureWidth() const { return mTextureWidth; }
+    uint32_t            GetFoveationTextureHeight() const { return mTextureHeight; }
+    uint32_t            GetFoveationTexelSize() const { return mTexelSize; }
+
+    uint32_t EncodeFragmentSizeFoveationTextureValue(uint8_t fragmentWidth, uint8_t fragmentHeight) const;
+    uint32_t EncodeFragmentDensityFoveationTextureValue(uint8_t xDensity, uint8_t yDensity) const;
+    void     FillFoveationTextureData(FoveationPatternType patternType, uint8_t* pData) const;
```

Comment:
Add FDM/VRS to function names?

Reply:
There are private FDM/VRS-specific versions of these functions. These functions allow code to specify either fragment size or density, and it will be encoded correctly for whichever foveation mode is being used.

Reply:
I've changed the way the shading rate images are filled, and moved this to a follow-up PR.